### DISCUSSION TITLE
Get the PID of _MY_  OpenConsole.exe?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#16447] [Q&A]
I think (could be wrong) that GetWindowThreadProcessId, used on the PseudoConsoleWindow class window, used to get the PID of OpenConsole.exe.  In any event, that doesn't seem to be true now; it provides the PID of the shell process.

Is there a way to determine the PID of the OpenConsole.exe which the shell is using?

Thanks!

**Comment by DHowett** [comment]
Honestly, I'm not sure -- we'll need to think about this one for a little bit.

What do you plan to do with this info? :)

**Comment by vefatica** [comment]
It's one of those "Am I running in Windows Terminal?" things.  **MY** OpenConsole.exe having WindowsTeminal as parent would be a good indicator.

I also noticed that `START /B & EXIT` now works to replace the shell instance without affecting WT or OpenConsole (a few years ago it did not work).  Is there any discussion of that change?  And I noticed that when I do that (even many times) GetWindowThreadProcessId continues to give the PID of the **FIRST** shell to use the PseudoConsoleWindow ... long after it terminated.  That seems a little bogus.

**Comment by 237dmitry** [comment]
> It's one of those "Am I running in Windows Terminal?" things. MY OpenConsole.exe having WindowsTeminal as parent would be a good indicator.

You can determine which shell running on which OpenConsole.exe

```powershell
$groups = Get-Process | Where-Object { $_.Parent.ProcessName -eq 'WindowsTerminal' }
                      | Group-Object { $_.StartTime.ToString() }

foreach ($group in $groups) { $group.Group; ''}
```

![Screenshot 2023-12-09 064619](https://github.com/microsoft/terminal/assets/78153320/5f261c8a-eebd-4af7-aefe-7b5deac376d7)

**Comment by vefatica** [comment]
I want my shell to get the ProcessId of the OpenConsole.exe in which it's running.  I can use C/C++ and the WIn32 API.

**Comment by vefatica** [comment]
Is there any more on this?

**Reply by lhecker** [reply]
<details>
<summary>Ah woops, damn...</summary>

```cpp
const auto GetRealWindowOwner = (DWORD(WINAPI*)(HWND))GetProcAddress(GetModuleHandleW(L"user32.dll"), MAKEINTRESOURCEA(2712));
```

![limesguy](https://github.com/microsoft/terminal/assets/2256941/cf3767aa-9957-483c-99d9-9befdc34f885)

</details>
